server/src/language-service/codeActionProvider.ts

```diff
@@ -0,0 +1,156 @@
+    public async provideCodeAction(
+        params: CodeActionParams,
+        doc: ITextDocument
+    ): Promise<CodeActionInsertEvr[]> {
+        let cnfg = await this.cnfgProvider.get(params.textDocument.uri);
+        if (!cnfg) {
+            return [];
+        }
+        let settings = await this.settingsManager.getSettings();
+        let insertPos = settings?.codeActions.insertEvrPosition ?? "bottom";
+        return getCodeActionInsertEvr(params, cnfg, doc, insertPos);
+    }
+}
+
+export interface CodeActionInsertEvr {
+    name: string;
+    uri: string;
+    offsetEdit: number;
+    diag: Diagnostic;
+}
+
+export function getCodeActionInsertEvr(
+    params: CodeActionParams,
+    cnfg: SepticCnfg,
+    doc: ITextDocument,
+    insertEvrPos: "top" | "bottom"
+): CodeActionInsertEvr[] {
+    let diag = getReferenceError(params.context.diagnostics);
+    if (!diag) {
+        return [];
+    }
+    let currentObject = cnfg.getObjectFromOffset(
+        doc.offsetAt(params.range.start)
+    );
+    if (!currentObject?.isType("CalcPvr")) {
+        return [];
+    }
+    let parent = currentObject.parent;
+    if (!parent?.isType("CalcModl")) {
+        return [];
+    }
+    while (parent && !parent.isType("DmmyAppl")) {
+        parent = parent.parent;
+    }
+    if (!parent) {
+        return [];
+    }
+    let objectBeforeInsert = parent;
+    if (insertEvrPos === "bottom") {
+        for (const child of parent.children) {
+            if (child.isType("CalcModl")) {
+                break;
+            }
+            objectBeforeInsert = child;
+        }
+    }
+    let referencedVariable = cnfg.getXvrRefFromOffset(
+        doc.offsetAt(params.range.start)
+    );
+    if (!referencedVariable) {
+        return [];
+    }
+    return [
+        {
+            name: referencedVariable.identifier,
+            uri: objectBeforeInsert.uri,
+            offsetEdit: objectBeforeInsert.end,
+            diag: diag,
+        },
+    ];
+}
+
+function getReferenceError(diagnostics: Diagnostic[]): Diagnostic | undefined {
+    for (const diag of diagnostics) {
+        if (diag.code === DiagnosticCode.missingReference) {
+            return diag;
+        }
+    }
+    return undefined;
+}
+
+export function createCodeActionInsertEvr(
+    action: CodeActionInsertEvr,
+    doc: ITextDocument
+): CodeAction {
+    let codeAction = CodeAction.create("Insert Evr");
+    codeAction.diagnostics = [action.diag];
+    codeAction.isPreferred = true;
+    codeAction.kind = CodeActionKind.QuickFix;
+    codeAction.edit = createInsertEvrEdit(action.name, action.offsetEdit, doc);
+    return codeAction;
+}
+
+function createInsertEvrEdit(
+    name: string,
+    offset: number,
+    doc: ITextDocument
+): WorkspaceEdit {
+    let editBuilder = new WorkspaceEditBuilder();
+    let text = getTextEvr(name);
+    editBuilder.insert(doc.uri, doc.positionAt(offset), text);
+    return editBuilder.getEdit();
+}
+
+function getTextEvr(name: string) {
```

Comment:
Possible to reuse snippet to keep code DRY?

Reply:
I don't think it is possible to reuse the snippet directly. The current version is copy paste from the snippet with the snippet specific things removed.